### PR TITLE
Add basic integration testing

### PR DESCRIPTION
**Pull request by reneradoi**
This PR adds a basic integration test to the etcd operator which:
- deploys a cluster of three etcd nodes
- checks the cluster's availability
- checks if all deployed nodes where added as cluster members
- writes test data to etcd
- reads the test data

### CODE REVIEW

File: tests/integration/helpers.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    """Resolve the etcd endpoints for a given juju application."""
+    return ",".join(
+        [
+            f"http://{unit.public_address}:{CLIENT_PORT}"
+            for unit in ops_test.model.applications[app_name].units
+        ]
+    )
+
+
+async def get_juju_leader_unit_name(ops_test: OpsTest, app_name: str = APP_NAME) -> str:
```
Comment (skourta) [inline tests/integration/helpers.py:57]: The return should be `str | None`. I know there is always a leader but str is not returned in all code paths.
You can also add a raise an error after the for loop.

File: tests/integration/helpers.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    return json.loads(result)["members"]
+
+
+def get_cluster_endpoints(ops_test: OpsTest, app_name: str = APP_NAME) -> str:
+    """Resolve the etcd endpoints for a given juju application."""
+    return ",".join(
+        [
+            f"http://{unit.public_address}:{CLIENT_PORT}"
+            for unit in ops_test.model.applications[app_name].units
```
Comment (skourta) [inline tests/integration/helpers.py:52]: I would add some asserts or tests on the model and the app to ensure they exist.

File: tests/integration/test_charm.py [hunk 3]
```diff
@@ -3,29 +3,49 @@
...
+    leader_unit = await get_juju_leader_unit_name(ops_test, APP_NAME)
+
+    cluster_members = get_cluster_members(model, leader_unit, endpoints)
+    assert len(cluster_members) == NUM_UNITS
+
+    # make sure data can be written to the cluster
+    test_key = "test_key"
+    test_value = "42"
+    assert put_key(model, leader_unit, endpoints, key=test_key, value=test_value) == "OK"
+    assert get_key(model, leader_unit, endpoints, key=test_key) == test_value
```
Comment (skourta) [inline tests/integration/test_charm.py:51]: I would check all 3 units separately to ensure it was replicated.

**Review by skourta — APPROVED** [review]
Thank you, René. This should get us started with the integration tests. I left enhancement comments but it already LGTM.

File: tests/integration/test_charm.py (hunk 3 above)
Comment (Mehdi-Bendriss) [reply tests/integration/test_charm.py:51]: May be premature, and will be addressed more in-depth during HA - but that's a valid point. 
For HA, we also need to have continuous writes routines implemented as part of the tests. Which should have replication checks happen as part of the continuous writes interface.

**Review by Mehdi-Bendriss — APPROVED** [review]

File: tests/integration/test_charm.py (hunk 3 above)
Comment (reneradoi) [reply tests/integration/test_charm.py:51]: Agree and will be done when working on HA.

**Comment by reneradoi** [comment]
Thank you for your feedback. We can address the comments when we further enhance the tests. Of course we need to add more tests and more functionality to the testing suite. This PR here is just considered a basic coverage to get fast feedback when making changes to the code, and to give a starting point on how to create the tests.

**Pull request merged**